### DISCUSSION TITLE
Exclude ranges from curve

### DISCUSSION THREAD

**Discussion by Troglobitten** [Rem0o/FanControl.Releases#1854] [Ideas]
**The problem**
Fans can become louder at specific speeds due to resonance. The only way I'm aware of to bypass this is to make a stairstep curve. While this does the job, it's very tedious, specially if you have to do this for individual fans. It also makes it hard to adjust your curves afterwards. 

**Suggestion**
My suggestion is to create an exclude ranges curve. With an option to select a baseline curve and to add exclusion ranges. If the speeds from the source curve are within range of the exclusion ranges, it would calculate the delta between the high and low value of the exclusion and jump to whichever is closest. 

[Quick Mockup](https://i.imgur.com/zIjMHjF.png)

**Comment by Rem0o** [comment]
Nice mockup! 

A very niche feature.   Question, how would that be less tedious than making a graph with the right steps for individual fans?
 With this new exclude function, you stil need to create a new curve per fan if the individual fans have different ranges to be "excluded".

**Reply by Troglobitten** [reply]
Yeah, you could argue that and you're not wrong. It just aligns with a modular/programmatic workflow where you have the freedom afterwards to adjust your base fan curve and it automatically applies to all the others. Like having a function in programming. You can call the function in your code, and still edit the function itself.

In my current setup I have 1 graph fan curve that gets modified by mix and sync curves to adjust individual fans. With an exclude range node I can do what you describe, and still modify the base curve easily. Without it, I will have to create separate graph curves and adjust each one manually for every change. 

Realistically speaking, users will not adjust their base curve every day. It'll be a useful feature for when fan wear adjust the acoustics of the fan, or for when users buy new parts. Pop in the new part, write down a few percentage values and you are up and running.

**Reply by Rem0o** [reply]
I don't hate it.

**Reply by kklittich** [reply]
I think this could be useful on a per fan basis, especially used with the Auto curve feature. You could use the auto curve and the fan could skip over (assume always overshoot?) the resonance zones quickly.

**Comment by yarmak72** [comment]
Good idea. It might be helpful for some fans like mine Arctic P14 PWM PST, which has very annoying noise at some RPM's.

**Comment by Rem0o** [comment]
However, isn't this control/fan specific? Why not make it a control setting, just like start and stop%? To me if this feature come through, it would make a lot more sense this way.  Whatever fan curve you apply to a control/fan, the excluded ranges would be applied.

**Reply by Troglobitten** [reply]
that makes a lot of sense. Wouldn't that make the control UI messy if you have to add multiple ranges to a single fan?

**Reply by Rem0o** [reply]
This would definitely be a dialog window kind of feature.

**Comment by Nightblade** [comment]
I've thought about this before.  My (probably overly complicated) idea:

1. Put a microphone near your PC
2. App monitors the mic volume
3. App sweeps through the fan's speeds (like in a calibration run)
4. App correlates mic volume with speed
5. App uses this noise-profile to create an "exclude ranges curve" as per OP's description

Internally this might look like a list of RPM/% speeds with a "Q" factor defining the width of the area to avoid, similar to how some audio equalizers work.

**Reply by kelthor85** [reply]
That's ... actually an excellent idea.  I was also thinking about tuning a receiver while reading your post.  I noticed you didn't mention measuring from the listening position, do you suppose this mightn't be possible as the overall output volume of the fans isn't that high? Likewise, I supposed you'd need a remarkably low noise floor (and a very good mic) so your close-to-case measurement does make sense. I wonder if there's any open versions of PEQ software...

**Reply by Nightblade** [reply]
Thanks!  Yes, the optimal distance between the PC and the microphone would need to be determined.  Another option might be to attach the microphone (or some kind of kinetic sensor) directly to one of the side-panels of the PC.
